fdbserver/KeyValueStoreMemory.actor.cpp

```diff
@@ -652,40 +677,118 @@ class KeyValueStoreMemory : public IKeyValueStore, NonCopyable {
+					//	.detail("PreviousSnapshotEndLoc", self->previousSnapshotEnd)
+					//	.detail("ThisSnapshotEnd", thisSnapshotEnd)
+					//	.detail("Items", snapItems)
+					//	.detail("CommittedWrites", self->notifiedCommittedWriteBytes.get())
+					//	.detail("SnapshotSize", snapshotBytes);
+
+					ASSERT(thisSnapshotEnd >= self->currentSnapshotEnd);
+					self->previousSnapshotEnd = self->currentSnapshotEnd;
+					self->currentSnapshotEnd = thisSnapshotEnd;
+
+					if (++self->snapshotCount == 2) {
+						self->replaceContent = false;
+					}
 
-				snapshotBytes = 0;
+					snapItems = 0;
+					snapshotBytes = 0;
+					snapshotTotalWrittenBytes += OP_DISK_OVERHEAD;
 
-				snapshotTotalWrittenBytes += OP_DISK_OVERHEAD;
-			} else {
-				StringRef tempKey = next.getKey(self->reserved_buffer);
-				self->log_op(OpSnapshotItem, tempKey, next.getValue());
-				nextKey = tempKey;
-				nextKeyAfter = true;
-				snapItems++;
-				uint64_t opBytes = tempKey.size() + next.getValue().size() + OP_DISK_OVERHEAD;
-				snapshotBytes += opBytes;
-				snapshotTotalWrittenBytes += opBytes;
+					// If we're not stopping now, reset next
+					if(snapshotTotalWrittenBytes < self->notifiedCommittedWriteBytes.get()) {
+						next = self->data.begin();
+					}
+					else {
+						// Otherwise, save state for continuing after the next wait and stop
+						nextKey = Key();
+						nextKeyAfter = false;
+						break;
+					}
+
+				} else {
+					// destKey is whichever of the two last key buffers we should write to next.
+					Key &destKey = lastSnapshotKeyUsingA ? lastSnapshotKeyA : lastSnapshotKeyB;
+
+					// Get the key, using destKey as a temporary buffer if needed.
+					KeyRef tempKey = next.getKey(mutateString(destKey));
+					int opKeySize = tempKey.size();
+
+					// If tempKey did not use the start of destKey, then copy tempKey into destKey.
+					// It's technically possible for the source and dest to overlap but with the current container implementations that will not happen.
+					if(tempKey.begin() != destKey.begin()) {
+						memcpy(mutateString(destKey), tempKey.begin(), tempKey.size());
+					}
+
+					// Now, tempKey's bytes definitely exist in memory at destKey.begin() so update destKey's contents to be a proper KeyRef of the key.
+					// This intentionally leaves the Arena alone and doesn't copy anything into it.
+					destKey.contents() = KeyRef(destKey.begin(), tempKey.size());
+
+					// Get the common prefix between this key and the previous one, or 0 if there was no previous one.
+					int commonPrefix;
+					if(useDelta) {
+						commonPrefix = commonPrefixLength(lastSnapshotKeyA, lastSnapshotKeyB);
+					}
+					else {
+						commonPrefix = 0;
+						useDelta = true;
+					}
+
+					// If the common prefix is greater than 1, write a delta item.  It isn't worth doing for 0 or 1 bytes, it would merely add decode overhead (string copying).
+					if(commonPrefix > 1) {
+						// Cap the common prefix length to 255.  Sorry, ridiculously long keys!
```

Comment:
Didn't you have a space efficient varint somewhere?

Reply:
You might be thinking of CompressedInt, used in TDMetrics, which is space efficient given its design constraint of supporting positive and negative numbers and having a numerical ordering that is identical to the lexical ordering of the resulting variable length byte strings.  The format is `<unary_byte_length><value_bits>` padded out to the total byte size indicated by the unary length.  The unary length is either
- one or more 0's followed by a 1 for negative numbers OR
- one or more 1's followed by a 0 for positive numbers.

So for a 1 byte integer the range would be 0 to 63 because the unary length of one requires 2 bits to represent.  We could use a common scheme where 1 bit of each byte is the "there are more bytes" bit, in which case one byte would have a range of 0 to 127.

However, since we have over 240 unused values in the log op code, we could just make another op code for, say `OpSnapshotItemDeltaLarge` which uses a 16 bit integer, which would cover more than the maximum key size FDB currently allows.

Note that if we zoom out a bit, the log op space efficiency is pretty bad anyway as it uses a 1 byte op code, a 4 byte length for param1, and a 4 byte length for param2.  That might be worth fixing, but has way more compatibility issues than adding `OpSnapshotItemDelta`.